### PR TITLE
Stop bundling unnecessary copy of ASM 5

### PR DESCRIPTION
**Pull request by basil**
See stapler/stapler#225 and jenkinsci/scm-api-plugin#88.

Jenkins core currently bundles upstream ASM 9.1 (a transitive dependency of JNR). Jenkins core also currently bundles a repackaged version of ASM 5 (a transitive dependency of Stapler). Maintaining a repackaged version of ASM does not seem like a desirable use of time for the Jenkins project. In a Java 11+ world, the use of ASM 5 (from the Java 8 era) seems increasingly problematic. Moreover, plugins like Token Macro already expect core to provide an upstream version of ASM as demonstrated in jenkinsci/token-macro-plugin#70, which updated Token Macro's POM to read:

```xml
    <dependency>
      <groupId>org.parboiled</groupId>
      <artifactId>parboiled-java</artifactId>
      <version>1.3.1</version>
      <exclusions>
        <!-- asm is provided by Jenkins core -->
        <exclusion>
          <groupId>org.ow2.asm</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

This PR formalizes the current status quo by defining the ASM version provided by Jenkins core in the Jenkins core BOM. This way, we can manage updates to it carefully and ensure they are properly tested. It also consumes stapler/stapler#225, which switches Stapler away from the repackaged version of ASM 5 to the latest upstream ASM (the same version we already bundle in Jenkins core via JNR).

This means Jenkins core will no longer provide a repackaged version of ASM 5. Besides Stapler and Jenkins core itself, the only plugin using the repackaged version of ASM 5 was [SCM API](https://plugins.jenkins.io/scm-api/). I verified this with `usage-in-plugins` and also by searching for [usages in sources](https://github.com/search?ref=simplesearch&type=Code&q=user%3Ajenkinsci+%22org.kohsuke.asm5%22). I opened jenkinsci/scm-api-plugin#88 to cover making SCM API consume the upstream version of ASM rather than the repackaged version.

### Proposed changelog entries

Developer: Jenkins now provides a stable version of [ObjectWebASM](https://asm.ow2.io/) (currently 9.1) on the classpath.

### Proposed upgrade guidelines

Jenkins previously provided both a repackaged version of [ObjectWebASM](https://asm.ow2.io/) 5 and the upstream version, but now it no longer provides the repackaged version of ASM 5. Users of [SCM API](https://plugins.jenkins.io/scm-api/) must upgrade SCM API to the latest version prior to upgrading Jenkins. Plugins that use the repackaged version of ASM 5 must be updated to use the upstream version of ASM instead.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — COMMENTED** [review]
looks good but requires https://github.com/stapler/stapler/pull/225 and https://github.com/jenkinsci/scm-api-plugin/pull/88 to be released first

**Comment by basil** [comment]
Another possibility would be to amend this PR to also retain a direct dependency on the repackaged version of ASM 5 (not exposed in the BOM). Thus we would still be delivering the repackaged version of ASM 5 purely as a compatibility shim for older versions of SCM API. That would allow us to move forward with this transition without worrying about breaking SCM API. When enough users of SCM API have upgraded to a version of SCM API that uses the upstream version of ASM, we could then complete the transition by dropping the direct dependency on ASM 5 from core.

**Comment by daniel-beck** [comment]
On-holding for upstream changes.

**Comment by basil** [comment]
> Another possibility

Proposed this alternative in #5525. I have no strong preference regarding one approach or the other. Happy to go with the community consensus.

File: core/src/main/java/hudson/util/SubClassGenerator.java
Comment (jglick) [inline core/src/main/java/hudson/util/SubClassGenerator.java:41]: Or just delete this craziness, as it appears unused?

Comment (basil) [reply core/src/main/java/hudson/util/SubClassGenerator.java]: I marked this with `@Deprecated` in the conservative version of this PR (#5225) and deleted the class and corresponding test in this PR.

File: core/src/main/java/hudson/util/SubClassGenerator.java
Comment (jglick) [inline core/src/main/java/hudson/util/SubClassGenerator.java:45]: By the way did you check the usage-in-plugins tool? I have not found any usages in @jenkinsci nor @cloudbees according to GitHub text search, though that is not totally reliable. Some interesting things I stumbled across:

* https://github.com/yannrouillard/pkg-jenkins/blob/f63a94ed21bd1584ab3043c33a4b22ee05a31b7e/debian/patches/build/0015-use-stock-asm.patch
* https://github.com/jenkins-infra/unused-code-detector/blob/b14b5cbf2880178cc7b48cecb846b66530ea7c03/Output_example.html#L1223

Comment (basil) [reply core/src/main/java/hudson/util/SubClassGenerator.java:45]: Yup, I ran `usage-in-plugins` against `hudson/util/SubClassGenerator` for update centers https://jenkins-updates.cloudbees.com/update-center/envelope-core-oc/update-center.json?version=2.277.4.3 and https://jenkins-updates.cloudbees.com/update-center/envelope-core-mm/update-center.json?version=2.277.4.3; found no usages.

**Comment by basil** [comment]
This PR is still blocked on the release and wide adoption of jenkinsci/scm-api-plugin#88.

File: core/src/main/java/hudson/util/SubClassGenerator.java
Comment (jtnord) [reply core/src/main/java/hudson/util/SubClassGenerator.java:46]: but what about actual shaded ASM classes?

**Review by jtnord — COMMENTED** [review]
ASM evolves in a not compatible way which is why it has been historically shaded.
By just "using the bom to manage it" makes things worse.  you may well have compile time failures if you update the jenkins version, but you will not stop runtime breakage.

PLugins should not be using ASM from core (unless it is a shaded version, and even then given we do not treat it as an API library probably a bad idea).

required for example by the byte-code-comptability library, which by the looks of it was accidentally switched to a non shaded version (probably by accident).

Guice uses 5.0.3 that we package (unshaded ASM should probably be banned by enforcer)

**Review by olamy — APPROVED** [review]

**Comment by jglick** [comment]
> PLugins should not be using ASM from core

Agreed, but I think this ship has sailed?

**Comment by timja** [comment]
> > PLugins should not be using ASM from core
> 
> Agreed, but I think this ship has sailed?

Discussion has moved on a bit since this in https://groups.google.com/g/jenkinsci-dev/c/NhV_o6zxbzw/m/xnI5Y7rPAQAJ

**Comment by timja** [comment]
SCM API has now been released and we can start thinking about when we want to ship this

**Comment by jglick** [comment]
https://github.com/jenkinsci/scm-api-plugin/releases/tag/scm-api-2.6.5 FTR

**Review by jglick — APPROVED** [review]

**Comment by basil** [comment]
> SCM API has now been released and we can start thinking about when we want to ship this

I don't have a strong preference, but there's no rush to ship this, so why don't we wait until the next LTS cycle begins to ensure no issues are encountered with the new dependency and to give users time to upgrade.

**Comment by timja** [comment]
Time to ship this @basil?

**Comment by basil** [comment]
> Time to ship this @basil?

Yeah, I think now is a good time. It's the beginning of a new LTS cycle, and users will need to upgrade SCM API for Guava compatibility anyway.

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
